pkg/controller/persistentvolume/types.go

```diff
@@ -111,16 +111,27 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *api.PersistentVo
 		// return the exact pre-binding match, if found
 		unboundVolumes := []*api.PersistentVolume{}
 		for _, volume := range volumes {
-			// volume isn't currently bound or pre-bound.
-			if volume.Spec.ClaimRef == nil {
-				unboundVolumes = append(unboundVolumes, volume)
+			// see if PV's claimRef matches claim
+			if volume.Spec.ClaimRef != nil {
+				if volume.Spec.ClaimRef.Name == claim.Name && volume.Spec.ClaimRef.Namespace == claim.Namespace {
```

Comment:
The UID check here was quite important. You must not bind to a volume that is Released and is waiting to be deleted or recycled.

I think this could work:
```
if volume.Spec.ClaimRef.Name == claim.Name && volume.Spec.ClaimRef.Namespace == claim.Namespace && ( volume.Spec.ClaimRef.UID=="" || volume.Spec.ClaimRef.UID == claim.UID) {
```

I.e. ClaimRef.UID must either match or be empty.